content/articles/web-apps.md

```diff
@@ -0,0 +1,124 @@
+    - front end: e.g Vue.js or React which is loaded in the user's browser and provides the user interface and visualisations.
+
+We run all of these components within [docker](https://docs.docker.com/get-started/) containers to provide reliable portability between installation environments. Typically the front and back ends of the web applications will go into one container, and the research code and interface layer will go into another. We'll run a web proxy in a third container, and there may be additional containers for databases, task queues etc.
+
+
+# Pre-requisites
+
+In order to support this architecture, we require the research code to have the following features so that we can build a stable interface to it:
+
+- **Source-controlled in git**: we maintain the source code for our applications in git repositories, and we require research code to be available in git too, so that we can develop against it, apply automated tests to it within [continuous integration](https://docs.github.com/en/actions/automating-builds-and-tests/about-continuous-integration) and deploy it using docker.
+
+- **Versioned**: it's useful to be able to surface a version number in the application, of both the research and application code, so that we can always be sure what version is deployed. It's helpful if the research code can provide its own semantic version, but as a minimum we could use a git commit identifier.
+
+- **Data sets are available and clearly defined**: any data required for the application (data which is used by the research code, or has been generated by it) should be available to automated access for CI and deployment. This could be on a network drive, but it would be preferable if it was in something like a github release. Data should also be versioned in some way, and its format should be identical between versions, except for known format upgrades.
+
+- **Stable**: we need to be able to rely on the research code to have a known interface and to run reliably. It's ok for the interface to evolve, as long as we can prepare for that in advance by updating the interface layer, and merging those changes into the research code and interface layer at the same time (along with any changes which might be required in the web application too). That way we can have a working main branch in all repositories all the time. It may be appropriate for some interface changes to be managed as evolving metadata so that application code does not need to change e.g. when a parameter is added.
+
+- **Tested**: it's good if the research code can have at least a basic suite of tests which run under continuous integration - it's often difficult to test exact output values from research code (e.g. if it is stochastic), but it's useful to at least test that results are being successfully returned in the expected format.
```

Comment:
Having shown this to people, we might separate out the CI part so that it stands out more, otherwise they seem to skip over it.